### PR TITLE
[Compute] Add disk-encryption-set command group

### PR DESCRIPTION
**Pull request by qwordy**
Issue #10948

Added `disk-encryption-set create`, `disk-encryption-set show`, `disk-encryption-set list`, `disk-encryption-set delete`, `disk-encryption-set update`
Added `--encryption-type` and `--disk-encryption-set` to `disk create`
Added `--os-disk-encryption-set` and `--data-disk-encryption-sets` to `vm create`
Added `--os-disk-encryption-set` and `--data-disk-encryption-sets` to `vmss create`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 1]
```diff
@@ -838,3 +838,11 @@ def load_arguments(self, _):
         c.argument('analytics_query', options_list=['--analytics-query', '-q'], help="Query to execute over Log Analytics data.")
         c.argument('timespan', help="Timespan over which to query. Defaults to querying all available data.")
     # endregion
+
+    # region disk encryption set
+    with self.argument_context('disk-encryption-set create') as c:
+        c.argument('disk_encryption_set_name', name_arg_type, help='Name of disk encryption set.')
+        c.argument('key_url', help='URL pointing to a key or secret in KeyVault.')
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: is the key_url same with Swagger property name? it's confusing with help message.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 2]
```diff
@@ -2933,3 +2933,23 @@ def execute_query_for_vm(cmd, client, resource_group_name, vm_name, analytics_qu
                        'Please check the status of log analytics workpsace.')
     return client.query(workspace, QueryBody(query=analytics_query, timespan=timespan))
 # endregion
+
+
+# disk encryption set
+def create_disk_encryption_set(cmd, client, resource_group_name, disk_encryption_set_name,
+                               key_url, source_vault, location=None, tags=None):
+    from msrestazure.tools import resource_id, is_valid_resource_id
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: just curious, why not import at global level? this is quite common methods

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: > 
> 
> is the key_url same with Swagger property name? it's confusing with help message.

Yes.
e.g. https://vault4848.vault.azure.net/keys/key1/e3a5735dfa4d47b8879fa61cdd1cfb82

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2019-07-01/disk.json#L1432

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 2 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: > 
> 
> just curious, why not import at global level? this is quite common methods

Follow the convention of existing functions. We can import them at global level, which saves many lines of code.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 3]
```diff
@@ -838,3 +839,12 @@ def load_arguments(self, _):
         c.argument('analytics_query', options_list=['--analytics-query', '-q'], help="Query to execute over Log Analytics data.")
         c.argument('timespan', help="Timespan over which to query. Defaults to querying all available data.")
     # endregion
+
+    # region disk encryption set
+    with self.argument_context('disk-encryption-set') as c:
+        c.argument('disk_encryption_set_name', disk_encryption_set_name)
+        c.argument('key_url', help='URL pointing to a key or secret in KeyVault.')
+        c.argument('source_vault', help='Resource ID of the KeyVault containing the key or secret.')
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: per cli command coding guideline,  does this source_vault support name of vault?

Comment (yungezz) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: based on code,  name is supported, pls update help message.

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Thanks for finding this issue. I'll fix it in help.

Comment (yungezz) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: seems not fixed yet

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 4]
```diff
@@ -109,6 +110,8 @@ def load_arguments(self, _):
                 c.argument('hyper_v_generation', arg_type=hyper_v_gen_sku, help='The hypervisor generation of the Virtual Machine. Applicable to OS disks only.')
             else:
                 c.ignore('access_level', 'for_upload', 'hyper_v_generation')
+            c.argument('encryption_type', arg_type=get_enum_type(self.get_models('EncryptionType')), help='Encryption type.')
+            c.argument('disk_encryption_set', help='Name of Id of disk encryption set that is used to encrypt the disk.')
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: typo in help.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 5]
```diff
@@ -373,6 +383,9 @@ def _build_storage_profile():
         if disk_info['os'].get('writeAcceleratorEnabled') is not None:
             profile['osDisk']['writeAcceleratorEnabled'] = disk_info['os']['writeAcceleratorEnabled']
         data_disks = [v for k, v in disk_info.items() if k != 'os']
+        for i, data_disk in enumerate(data_disks):
+            if i < len(data_disk_encryption_sets):
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: what if data_disks len  not matching with len of data_disk_encryption_sets? better to add check

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Fixed

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 5 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: Fixed

File: src/azure-cli/azure/cli/command_modules/vm/_help.py [hunk 6]
```diff
@@ -100,6 +100,21 @@
     crafted: true
 """
 
+helps['disk-encryption-set'] = """
+type: group
+short-summary: Disk Encryption Set resource.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/_help.py]: @ramankumarlive, could you help give a summary of disk encryption set?

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Maybe the name `--key` or `--active-key` is better than `--key-url`.

File: src/azure-cli/azure/cli/command_modules/vm/commands.py [hunk 7]
```diff
@@ -172,6 +173,11 @@ def load_command_table(self, _):
         client_factory=cf_log_analytics_data_plane,
     )
 
+    compute_disk_encryption_set_sdk = CliCommandType(
+        operations_tmpl='azure.mgmt.compute.operations#DiskEncryptionSetsOperations.{}',
+        client_factory=cf_disk_encryption_set
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/commands.py]: actually, I'm curious about the api-version of this command type. Sometimes you need to set operation_group for the commandType in VM, sometimes you don't. Is there any guideline for this?

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 8]
```diff
@@ -2954,3 +3002,41 @@ def execute_query_for_vm(cmd, client, resource_group_name, vm_name, analytics_qu
...
+        return client.list_by_resource_group(resource_group_name)
+    return client.list()
+
+
+def update_disk_encryption_set(instance, client, resource_group_name, key_url=None, source_vault=None):
+    from msrestazure.tools import resource_id, is_valid_resource_id
+    if not is_valid_resource_id(source_vault):
+        source_vault = resource_id(subscription=client.config.subscription_id, resource_group=resource_group_name,
+                                   namespace='Microsoft.KeyVault', type='vaults', name=source_vault)
+    if key_url:
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py]: user cannot set `key_url` to none. right?

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: It's required.

File: src/azure-cli/azure/cli/command_modules/vm/commands.py (hunk 7 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/commands.py]: operation_group - Only used by the azure-cli-vm module to specify which resource API to target. No

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 8 above)
Comment (MyronFanQiu) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: `key_url=None`?

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py]: When key_url == None, it means do not update key_url. In a disk encryption set, key_url can not be None

**Pull request merged**